File: 71-simplify-path/simplify-path.java
class Solution {
    public String simplifyPath(String path) {
         Stack<String> st = new Stack<>();
        
        // String ko '/' ke basis par tod kar array bana liya
        String[] tokens = path.split("/");
        
        // For-each loop lagaya saare tukdon par
        for (String token : tokens) {
            // Agar khali hai ya '.' hai toh ignore karo
            if (token.equals("") || token.equals(".")) {
                continue;
            }
            // Agar '..' aaya toh ek kadam peeche hato
            if (token.equals("..")) {
                if (!st.isEmpty()) {
                    st.pop();
                }
            } 
            // Normal naam hai toh stack mein push karo
            else {
                st.push(token);
            }
        }
        
        // Stack se nikal kar wapas seedhi string banana
        StringBuilder ans = new StringBuilder();
        for (String dir : st) {
            ans.append("/").append(dir);
        }
        
        // Agar length 0 hai toh "/" de do, nahi toh bani hui string return karo
        return ans.length() == 0 ? "/" : ans.toString();
    }
}

File: 71-simplify-path/simplify-path.cpp
class Solution {
public:
    string simplifyPath(string path) {
        vector<string> st; // Hum vector ko hi stack ki tarah use kar rahe hain
        stringstream ss(path);
        string token;
        
        // Yeh loop string ko '/' ke basis par tod kar ek-ek tukda 'token' mein dega
        while (getline(ss, token, '/')) {
            // Agar tukda khali hai ya single dot '.' hai, toh kuch mat karo
            if (token == "" || token == ".") {
                continue;
            }
            // Agar double dot '..' aaya, toh ek kadam peeche jao (pop)
            if (token == "..") {
                if (!st.empty()) {
                    st.pop_back(); // Pichle folder ko remove kiya
                }
            } 
            // Agar koi valid naam hai (jaise 'home', '...', 'foo')
            else {
                st.push_back(token); // Stack mein folder ka naam daal do
            }
        }
        
        // Aakhiri mein saare bache hue folders ko '/' se jod kar final path banao
        string ans = "";
        for (string dir : st) {
            ans += "/" + dir;
        }
        
        // Agar stack bilkul khali tha, toh sirf "/" return karo, nahi toh ans return karo
        return ans.empty() ? "/" : ans;
    }
};